File: q2_ms/xcms/fill_peaks_area.py
import copy

from q2_ms.types import XCMSExperimentDirFmt, mzMLDirFmt
from q2_ms.utils import run_r_script


def fill_peaks_area(
    spectra: mzMLDirFmt,
    xcms_experiment: XCMSExperimentDirFmt,
    mz_min: str = "function(z) quantile(z, probs = 0.25)",
    mz_max: str = "function(z) quantile(z, probs = 0.75)",
    rt_min: str = "function(z) quantile(z, probs = 0.25)",
    rt_max: str = "function(z) quantile(z, probs = 0.75)",
    threads: int = 1,
    ms_level: int = 1,
) -> XCMSExperimentDirFmt:
    # Create parameters dict
    params = copy.copy(locals())

    # Innit XCMSExperimentDirFmt
    xcms_experiment = XCMSExperimentDirFmt()

    # Add output path to params
    params["output_path"] = str(xcms_experiment)

    # Run R script
    run_r_script(params, "fill_peaks_area", "XCMS")

    return xcms_experiment

File: q2_ms/assets/fill_peaks_area.R
#!/usr/local/bin/Rscript --vanilla

library(xcms)
library(MsExperiment)
library(MsIO)
library(Spectra)
library(optparse)

# Define command-line options
option_list <- list(
  make_option(opt_str = "--spectra", type = "character"),
  make_option(opt_str = "--xcms_experiment", type = "character"),
  make_option(opt_str = "--mz_min", type = "character"),
  make_option(opt_str = "--mz_max", type = "character"),
  make_option(opt_str = "--rt_min", type = "character"),
  make_option(opt_str = "--rt_max", type = "character"),
  make_option(opt_str = "--threads", type = "integer"),
  make_option(opt_str = "--output_path", type = "character"),
  make_option(opt_str = "--ms_level", type = "integer")
)

# Parse arguments
optParser <- OptionParser(option_list = option_list)
opt <- parse_args(optParser)

XCMSExperiment <- readMsObject(XcmsExperiment(), PlainTextParam(opt$xcms_experiment))

# Set parameters
AreaParams <- ChromPeakAreaParam(
  mzmin = eval(parse(text = opt$mz_min)),
  mzmax = eval(parse(text = opt$mz_max)),
  rtmin = eval(parse(text = opt$rt_min)),
  rtmax = eval(parse(text = opt$rt_max))
)

# Fill gaps
XCMSExperiment <- fillChromPeaks(
  object = XCMSExperiment,
  param = AreaParams,
  msLevel = opt$ms_level,
  BPPARAM = MulticoreParam(workers = opt$threads)
)

# Export the XCMSExperiment object to the directory format
saveMsObject(XCMSExperiment, param = PlainTextParam(path = opt$output_path))
